googletest/CMakeLists.txt

```diff
@@ -72,6 +72,25 @@ include(cmake/internal_utils.cmake)
 
 config_compiler_and_linker()  # Defined in internal_utils.cmake.
 
+# Create the CMake package file descriptors.
+if (INSTALL_GTEST)
+  include(CMakePackageConfigHelpers)
+  set(cmake_package_name GTest)
+  set(targets_export_name ${cmake_package_name}Targets CACHE INTERNAL "")
```

Comment:
I think it make sense just to move this code to top-level CMakeLists.txt, in this case there is no need to use `CACHE INTERNAL` variables. Also this file is optional, we will not reach this code if `BUILD_GTEST` is `OFF`.

Reply:
@ruslo, what you suggest has been my first attempt because it is logical. Though, I had to revert it to the current code because:
1. `googlemock/CMakeLists.txt` actually `add_subdirectory()` the `gtest` project. Therefore the above code fragment is always included, even when `BUILD_GTEST=OFF`.
2. To be able to inject the `GTEST_HAS_PTHREAD` into the generated `Config.cmake` file, this code shall be included _after_ the call to `config_compiler_and_linker()`, which is currently in `googletest/CMakeLists.txt`. At the same time it shall be _before_ the calls to our `install_project()`. Putting it in the top-level `CMakeLists.txt` does not satisfy either one or the other requirement.

WDYT?

Reply:
Afterthought: AFAIK `BUILD_GTEST` is vestigial from the time they were two separate projects, i.e. with different lifecycles. IMHO, the case `BUILD_GTEST=OFF AND BUILD_GMOCK=OFF` is not very interesting and could be simplified away by dropping `BUILD_GTEST` completely. This is an item for the list of next steps.

Reply:
> not very interesting and could be simplified away by dropping BUILD_GTEST completely

👍 

> googlemock/CMakeLists.txt actually add_subdirectory() the gtest project

> this code shall be included after the call to config_compiler_and_linker()

I think this code should be refactored.

Reply:
I removed `BUILD_GTEST`. Further reorganization is possible but the side effects are not trivial to follow. Let's discuss them later.

Reply:
> Let's discuss them later

From my opinion they should be fixed **before** this pull request proceed.

Reply:
@ruslo, I did some changes in this direction, including probably something too much. But let's start the discussion. I split the changes in more commits trying to keep functionally related stuff together. This grouping is subject to discussion too.

Side note: I notice that previously, two of the python tests were not detected by CMake, probably due to the strange inclusion order of the projects. These are `gmock_leak_test` and `gmock_output_test`.

Reply:
> I did some changes in this direction, including probably something too much

Yes, I don't think it is a good idea to mix everything into one big change. For review and for testing reasons it will be better to create several pull requests, e.g. "Set CMake 3.1 as required minimum".